### ISSUE TITLE
[video_player] accept stream as input

### ISSUE THREAD

**Issue by neckaros** [flutter/flutter#49967]
## Use case

I obtain video data behind protected API (ex google API) and I want to play them in the video player. Sometime I also want to decrypt them if the user chose to encrypt it with his own key for privacy 

## Proposal

What I think of doing is creating an http server and give local address to the network constructor but it seems like too much overhead and not completely secure.

It would be great if we could pipe a stream or even have a kind of callback system to allow fetching of specific range to allow seeking

**Comment by VladyslavBondarenko** [comment]
Related to https://github.com/flutter/flutter/issues/24960

**Comment by cachapa** [comment]
Similar use case and need: our videos are encrypted end-to-end which means they need to be decrypted locally before being fed to video_player.
We already support streamed decryption, but found no way to feed the plaintext end of that stream to the video controller.

**Comment by theskyblockman** [comment]
I've looked into this but because of the way the data sources are given to the native part of the package, it would require a big rewrite of how the package gives data to be rendered. So if I want to use the video player on Android over a local file, the Dart package part of the package will give to its native part only the file path and would read the file from there resulting of the use of only 1 communication from the Dart part of the package to its native one. But if a Stream is used, it would require constant communication between the two parts of the package and the way the package is written couldn't support this. That's why I think a bigger pull request should be made to do a big rewrite of this library to be more modern, and if it happens at any time, I would be pleased to contribute to the rewrite myself.

TL;DR: This is pretty much impossible to implement this feature without a package rewrite

**Comment by rohintoncollins** [comment]
I also need this. I am downloading chunks of an encrypted file from a private server. I then have a transformer which decrypts the chunks. This is all Stream<List<int>>. I need a media player which will accept a Stream<Lint<int>>. I haven't managed to find anything yet. I would have thought, that given that Files are streams, or at least can be converted to streams (e.g. (File().openRead() returns a stream) that there would be a solution somewhere for packages which use files. My app is initially a Web App, so I am not using a file system.

**Comment by theskyblockman** [comment]
> My app is initially a Web App, so I am not using a file system.

@rohintoncollins I would advise you to look at [this](https://stackoverflow.com/a/76312355/14406240). As you add decryption this could not be applicable, take a look at [VideoPlayerController.networkUrl](https://pub.dev/documentation/video_player/latest/video_player/VideoPlayerController/VideoPlayerController.networkUrl.html) and the headers part to see if you couldn't instead provide your key to your server which will then let you stream the raw data directly to the native side.

For other platforms than web this couldn't be possible at all.

**Comment by rohintoncollins** [comment]
@theskyblockman Thanks for your quick reply. Yes I saw the [link](https://stackoverflow.com/a/76312355/14406240) you posted, but could not get this to work. In any case I think I now understand the issue. For anyone else who finds themselves in the same situation I am in: trying to play a video from an encrypted data source.
  Video players support random-access data sources of unencrypted data. This is generally a file or a URL which supports ranges. The random access nature of the data source enables scrubbing (advance/rewind). The video player has codecs, but won't support decryption. Generally encryption/decryption is serial in nature (i.e. not random-access). You could create a stream of downloaded encrypted chunk -> decrypted chunk -> media player, but this is not random access, and video players do not support streams for this reason. In order to make the decryption random access you would need to individually encrypt each chunk. In addition you would need to build the decryption logic into the video player - just when it fetches an encrypted chunk from a random access data source. I am building a Flutter Web App, so do not have access to the file system. I do have a random access URL, but only of encrypted data where encryption is at the file level. So I will need to:
  - switch to using block-based encryption (say of 2MB blocks)
  - fork video_player or media_player and add my decryption logic at the place where a block is downloaded.

**Comment by neckaros** [comment]
You could use AES CBC for block encryption, you just have to get the previous 16Bits to get the IV and then you can decode the requested range

**Comment by ElvinSadikhov** [comment]
hello. are there any news on this matter? I also need to implement this feature

**Comment by helomri** [comment]
> hello. are there any news on this matter? I also need to implement this feature

It _is_ in my backlog which means I will probably look at it at some point in the future, but seeing what I'm working on right now, it could take me a _really_ long time to get there, except if someone else does it. The Flutter team has bigger fish to fry right now (the chance they do it is probably 0%.) I would rather count on a third-party implementing this in a package and posting a link to it in this discussion.

To put it in simple terms: it ain't happening anytime soon.
